micrometer-core/src/main/java/io/micrometer/core/instrument/binder/commonspool2/CommonsObjectPool2Metrics.java

```diff
@@ -0,0 +1,331 @@
+import io.micrometer.core.lang.Nullable;
+import io.micrometer.core.util.internal.logging.InternalLogger;
+import io.micrometer.core.util.internal.logging.InternalLoggerFactory;
+
+import javax.management.AttributeNotFoundException;
+import javax.management.InstanceNotFoundException;
+import javax.management.ListenerNotFoundException;
+import javax.management.MBeanException;
+import javax.management.MBeanServer;
+import javax.management.MBeanServerDelegate;
+import javax.management.MBeanServerFactory;
+import javax.management.MBeanServerNotification;
+import javax.management.MalformedObjectNameException;
+import javax.management.NotificationFilter;
+import javax.management.NotificationListener;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+import java.lang.management.ManagementFactory;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.ToDoubleFunction;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * Apache Commons Pool 2.x metrics collected from metrics exposed via the MBeanServer.
+ * Metrics are exposed at each object pool.
+ *
+ * @author Chao Chang
+ */
+public class CommonsObjectPool2Metrics implements MeterBinder, AutoCloseable {
+    private final static InternalLogger log = InternalLoggerFactory.getInstance(CommonsObjectPool2Metrics.class);
+    private static final String JMX_DOMAIN = "org.apache.commons.pool2";
+    private static final String METRIC_NAME_PREFIX = "commons.pool2.";
+
+    private static final String[] TYPES = new String[]{"GenericObjectPool", "GenericKeyedObjectPool"};
+
+    private final Executor executor = Executors.newSingleThreadExecutor();
+
+    private final MBeanServer mBeanServer;
+    private final Iterable<Tag> tags;
+    private final List<Runnable> notificationListenerCleanUpRunnables = new CopyOnWriteArrayList<>();
+
+    public CommonsObjectPool2Metrics() {
+        this(emptyList());
+    }
+
+    public CommonsObjectPool2Metrics(Iterable<Tag> tags) {
+        this(getMBeanServer(), tags);
+    }
+
+    public CommonsObjectPool2Metrics(MBeanServer mBeanServer, Iterable<Tag> tags) {
+        this.mBeanServer = mBeanServer;
+        this.tags = tags;
+    }
+
+    private static MBeanServer getMBeanServer() {
+        List<MBeanServer> mBeanServers = MBeanServerFactory.findMBeanServer(null);
+        if (!mBeanServers.isEmpty()) {
+            return mBeanServers.get(0);
+        }
+        return ManagementFactory.getPlatformMBeanServer();
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        for (String type : TYPES) {
+            registerMetricsEventually(
```

Comment:
Is there a non-JMX way to register these?

Reply:
Other ways are not always possible as far as I know, because libraries using commons-pool2(like jedis or dbcp2) keeps the pool objects as internal properties, and we cannot get them.
and commons-pool exposes JMX, so using JMX seems to be a natural fit.

Reply:
I agree JMX as the default is a good idea. I was more curious if there was a way if we did have access to the pool directly. I used commons pool in some of my own code and didn't have JMX enabled at the time. It would be nice to offer a way to still gather those metrics if posssible.

Reply:
As we can see from the following code, JMX registration in commons-pool2 is enabled by default , unless you configure to disable it explicitly, which I don't think a serious application should do.
https://github.com/apache/commons-pool/blob/2d2049233940b10c7dd35df70a353d383f601053/src/main/java/org/apache/commons/pool2/impl/BaseObjectPoolConfig.java#L139